### ISSUE TITLE
Debug console print() statements return printed object + None

### ISSUE THREAD

**Issue by ReveStobinson** [microsoft/debugpy#678]
## Environment data

-   **VS Code version:** 1.45.1
-   **Extension version:** 2020.6.91350
-   **OS and version:** Windows 10 Enterprise, version 1909, Build 18363.1679
-   **Python version (& distribution if applicable, e.g. Anaconda):** Python 3.7.0
-   **Type of virtual environment used:** virtualenv
-   **Relevant/affected Python packages and their versions:** N/A
-   **Relevant/affected Python-related VS Code extensions and their versions:** N/A
-   **Value of the `python.languageServer` setting:** Jedi

## Expected behaviour

Calling `print(x)` in debug console with `"redirectOutput": true` in launch.json prints `x` to the debug console

## Actual behaviour

Behavior instead looks like this:
```
>>>print("Hello world!")
Hello world!
None
```

## Notes

This is occurring after I found and implemented the fix for https://github.com/microsoft/ptvsd/issues/2036. Before, any print statement always just returned `None`. After adding `"redirectOutput": true` to my launch.json, print statements print the desired object, but the `None` still remains on the next line.

Changing `"console":` from `"integratedTerminal"` to `"integratedConsole"` in launch.json also seems to have no effect.

**Comment by fabioz** [comment]
From what I can tell, the `"redirectOutput": true` isn't needed (the extension version: `2020.6.91350` seems to be pretty old -- please try the one with the latest debugger (`2021.7.1060902895`)).

Now, `None` is still printed in the end. 

Right now this is by design: if the code can be evaluated as a single statement and it has a return, that return is shown (whether it's `None` or not).

Note that it's shown with a different color from the output color (which VSCode uses to signal that this is the result of the evaluation).

We could probably do a special-casing just for the `print` function and don't provide a return for VSCode in this case, but I'm not sure if it's worth it or not.

Opinions @karthiknadig @int19h?

Also, I'm curious, @ReveStobinson why are you using `print` at all (in general, you could just type in the variable you want and it'd be printed -- calling the print seems to be redundant for me).

**Comment by int19h** [comment]
The standard Python REPL doesn't output `None` when the expression produces that; we should probably just do the same. I'm not entirely sure how to wire it up to VSCode, though. Perhaps if we see the evaluation context equal to "repl", report the expression as not returning a value at all if it returned `None`?

**Comment by luabud** [comment]
I like Pavel's point about matching the same behaviour as the Python REPL 😊

**Comment by pabloazurduy** [comment]
Hi everyone, 
I have a similar issue as the one described in #88362, I've tried everything, but `print('str')` still don't appear anywhere. I have the following configuration `launch.json` for some reason I've tried all the configurations possible:
with `"redirectOutput": true`with `"console": "internalConsole",` also with `integratedTerminal` and so on, I have another mac that shows the prints even using the same configuration. I seriously don't know what's going on  
 ```json
   // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "env": {//"DISPLAY":":1",
                    //"PYTHONPATH": "${workspaceRoot}",
                    //"PYDEVD_WARN_EVALUATION_TIMEOUT": "500",
                    //"PYDEVD_UNBLOCK_THREADS_TIMEOUT": "500",
                    //"PYDEVD_THREAD_DUMP_ON_WARN_EVALUATION_TIMEOUT":"true"
                },
            "cwd": "${workspaceFolder}",
            "console": "internalConsole",
            "redirectOutput": true
        }
    ]
}


```
Any ideas ?

**Comment by fabioz** [comment]
Can you provide the logs for the run?

i.e.:
* Run the command `Extensions: Open Extensions Folder`
* Locate the Python extension directory, typically of the form `ms-python.python-2022.*.****`
* In that directory ensure you do not have any debug*.log files, if you do, please delete them
* Go back into VS Code and modify your launch.json to add the setting `"logToFile": true`, see below:

```
{
    "name": "Python: Current File (Integrated Terminal)",
    "type": "python",
    "request": "launch",
    "program": "${file}",
    "console": "integratedTerminal",
    "logToFile": true
}
```

* Start debugging
* When done, go back into the extension directory and upload the debug*.log files into this GitHub issue.

**Comment by pabloazurduy** [comment]
hi @fabioz 
I tried using the configuration that you gave me 
<img width="691" alt="image" src="https://user-images.githubusercontent.com/24685386/163442869-1654bcb3-c5f1-4c44-b625-a022309386d7.png">
but for some reason the `degug.log` file never appeared in the extension folder 
<img width="436" alt="image" src="https://user-images.githubusercontent.com/24685386/163442973-c1c056f7-34a2-472c-8673-acf59ba57c84.png">
do you know what might be happening ?

**Comment by fabioz** [comment]
I believe it's not picking up that extension when you're launching (you're probably launching through the dropdown). To launch using the launch configuration you need to run it in the debug view.

i.e.:

![image](https://user-images.githubusercontent.com/117621/163443593-d194cd65-5b2b-4998-99b5-967a0e8ff805.png)

**Comment by pabloazurduy** [comment]
ohhh I'm running it via the `TESTING` tab using the debug button, the thing is I'm running tests (via-pytest) 

<img width="273" alt="image" src="https://user-images.githubusercontent.com/24685386/163443938-af69fb9f-8a1e-482c-8dc7-90b9bf260d1e.png">
 adding    `"purpose": ["debug-test"], `might solve that issue ?

**Comment by karthiknadig** [comment]
@pabloazurduy That is correct. Adding `"purpose": ["debug-test"], ` will mark the configuration to use when run via test explorer.

**Comment by pabloazurduy** [comment]
Another update: I've tried with the configuration:
```json
        {
            "name": "Python: Current File (Integrated Terminal)",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "logToFile": true,
            "purpose": ["debug-test"], 
            "redirectOutput": true,
            "env": {"PYTHONPATH": "${workspaceRoot}"}
        }
```
but instead of launching from the *testing tab* I used the *Run and debug tab* and I launched the module that I was testing adding a main, using a break point worked

```python 
# somecode 
if __name__ == '__main__':
    print('something')
```
<img width="408" alt="image" src="https://user-images.githubusercontent.com/24685386/163460598-d76dabfc-4b12-4dd5-bd09-5fe03f5e8362.png">

still is for some reason this is not working when launching the pytest from the *testing tab*, I should be using the same config no ?. I don't understand where the problem is coming from.

**Comment by fabioz** [comment]
@pabloazurduy this is because `pytest` captures the output.

See: https://github.com/microsoft/debugpy/issues/899#issuecomment-1095384631

**Comment by pabloazurduy** [comment]
Thanks @fabioz ! that was it !!!! 
adding 
`"python.testing.pytestArgs": ["--capture=no"],` into the `settings.json` solved the issue ! 
many thanks !

**Comment by avraam-inside** [comment]
> Спасибо,@fabioz! это было!!!! (Отредактировано) Я предложил изменить параметры с vscode таким образом `pytest`, чтобы он не перенаправлялся в сторону, очень раздражает отсутствие `print`работы оператора, особенно когда вы хотите отлаживать некоторые объекты и сценарии, которые имеют стандартный вывод или итерацию и печать.
> 
> в случае возникновения, как@fabiozупоминается в комментариях [# 899](https://github.com/microsoft/debugpy/issues/899) , это происходит с тех пор, как последняя версия `debugpy v1.6.0`автоматически обновляется vscode, поскольку обходной путь, который сработал для меня, заключен в том, чтобы понизить версию расширения python для той, которая была до того, как эта функция была решена в консоли отладки .[`2022.4.0`](https://github.com/microsoft/vscode-python/releases/tag/2022.4.0)`v2022.2.1924087327``print()`
> 
> это также было причиной, по которой он работал на моем компьютере Mac, отладочная версия была`v1.5.1`
> 
> ### tldr
> Понизьте расширение vscode python до `v2022.2.1924087327`того, что поможет

Thanks! It works!

I hope that the extension developers will fix this problem in the next versions

**Comment by stev-0** [comment]
Although it's a bit of a faff, you can also add the following to your `settings.json`. This is global though so you need to change it each time you want to change the output.
   ```
 "python.testing.pytestArgs": [
        "-s",
    ]
```
